core/include/gnuradio-4.0/Block.hpp

```diff
@@ -644,6 +653,30 @@ struct Block : protected std::tuple<Arguments...> {
     friend constexpr auto &
     outputPort(Self *self) noexcept;
 
+    constexpr void
+    checkParametersAndThrowIfNeeded() {
+        constexpr bool kIsSourceBlock = traits::block::input_port_types<Derived>::size == 0;
+        constexpr bool kIsSinkBlock   = traits::block::output_port_types<Derived>::size == 0;
+
+        if constexpr (Resampling::kEnabled) {
+            static_assert(!kIsSinkBlock, "Decimation/interpolation is not available for sink blocks. Remove 'ResamplingRatio<>' from the block definition.");
+            static_assert(!kIsSourceBlock, "Decimation/interpolation is not available for source blocks. Remove 'ResamplingRatio<>' from the block definition.");
+            static_assert(HasProcessBulkFunction<Derived>, "Blocks which allow decimation/interpolation must implement processBulk(...) method. Remove 'ResamplingRatio<>' from the block definition.");
+        } else {
+            if (numerator != 1_UZ || denominator != 1_UZ) {
+                throw std::runtime_error(fmt::format("Block is not defined as `ResamplingRatio<>`, but numerator = {}, denominator = {}, they both must equal to 1.", numerator, denominator));
+            }
+        }
+
+        if constexpr (StrideControl::kEnabled) {
+            static_assert(!kIsSourceBlock, "Stride is not available for source blocks. Remove 'Stride<>' from the block definition.");
+        } else {
+            if (stride != 0_UZ) {
```

Comment:
TODO: should constrain `stride` to be non-zero using the `Annotated<>` interface. This way this is only checked when stride is changed and not during each iteration.